src/snowflake/snowpark/mock/_functions.py

```diff
@@ -304,34 +306,66 @@ def mock_to_date(
 
         [x] If the variant contains a string, a string conversion is performed.
 
-        [ ] If the variant contains a date, the date value is preserved as is.
+        [x] If the variant contains a date, the date value is preserved as is.
 
-        [ ] If the variant contains a JSON null value, the output is NULL.
+        [x] If the variant contains a JSON null value, the output is NULL.
 
         [x] For NULL input, the output is NULL.
 
-        [ ] For all other values, a conversion error is generated.
+        [x] For all other values, a conversion error is generated.
     """
     res = []
-    auto_detect = bool(not fmt)
 
-    date_format, _, _ = convert_snowflake_datetime_format(
-        fmt, default_format="%Y-%m-%d"
-    )
+    if not isinstance(fmt, ColumnEmulator):
+        fmt = [fmt] * len(column)
+
+    for data, _fmt in zip(column, fmt):
+        auto_detect = _fmt is None or _fmt.lower() == "auto"
+
+        date_format, _, _ = convert_snowflake_datetime_format(
+            _fmt, default_format="%Y-%m-%d"
+        )
+        import dateutil.parser
 
-    for data in column:
         if data is None:
             res.append(None)
             continue
         try:
-            if auto_detect and data.isnumeric():
-                res.append(
-                    datetime.datetime.utcfromtimestamp(
-                        convert_integer_value_to_seconds(data)
-                    ).date()
-                )
+            if isinstance(column.sf_type.datatype, TimestampType):
+                res.append(data.date())
+            elif isinstance(column.sf_type.datatype, StringType):
+                if data.isdigit():
+                    res.append(
+                        datetime.datetime.utcfromtimestamp(
+                            convert_integer_value_to_seconds(data)
+                        ).date()
+                    )
+                else:
+                    if auto_detect:
+                        res.append(dateutil.parser.parse(data).date())
+                    else:
+                        res.append(datetime.datetime.strptime(data, date_format).date())
+            elif isinstance(column.sf_type.datatype, VariantType):
+                if isinstance(data, str):
+                    if data.isdigit():
+                        res.append(
+                            datetime.datetime.utcfromtimestamp(
+                                convert_integer_value_to_seconds(data)
+                            ).date()
+                        )
+                    else:
+                        # for variant type with string value, snowflake auto-detects the format
```

Comment:
Question: `fmt` will be ignored even if specified?

Reply:
I checked, error will be raised instead of being ignored.
I updated the code